### DISCUSSION TITLE
The type initializer for 'X64' threw an exception.

### DISCUSSION THREAD

**Discussion by nsutariya9393** [dlemstra/Magick.NET#874] [Help]
hello, 

I am using magick.net to convert psd to png images for my web application project with .net framework 4.0 and works perfectly on my local machine but when I deploy on server, it started giving error "The type initializer for 'X64' threw an exception." I have tried all the solutions I found but none of them worked for me. The application is running on x64. I have tried adding 64 and anycpu dlls for magick.net.

can anyone please help me? I am stuck from a week.

**Comment by dlemstra** [comment]
Is it possible to get the complete stack trace of the exception?

**Reply by nsutariya9393** [reply]
[DllNotFoundException: Unable to load DLL 'Magick.Native-Q16-x64.dll': The specified procedure could not be found. (Exception from HRESULT: 0x8007007F)]
   ImageMagick.X64.Environment_SetEnv(IntPtr name, IntPtr value) +0
   ImageMagick.NativeEnvironment.SetEnv(String name, String value) +150
   ImageMagick.MagickNET.InitializePrivate(ConfigurationFiles configFiles, String newPath) +39
   ImageMagick.X64..cctor() +15

[TypeInitializationException: The type initializer for 'X64' threw an exception.]
   ImageMagick.X64.Environment_SetEnv(IntPtr name, IntPtr value) +0
   ImageMagick.NativeEnvironment.SetEnv(String name, String value) +150
   ImageMagick.MagickNET.SetTempDirectory(String path) +30
   PINT.btnSaveSKUPrintImages_Click(Object sender, EventArgs e) in c:\inetpub\wwwroot\test.aspx.cs:736
   System.Web.UI.WebControls.Button.OnClick(EventArgs e) +11750641
   System.Web.UI.WebControls.Button.RaisePostBackEvent(String eventArgument) +150
   System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint) +6016

**Reply by nsutariya9393** [reply]
only thing I want to know is, where should be native dll on server? it is there already in temp folder during runtime. but still it couldn't find.

**Reply by nsutariya9393** [reply]
@dlemstra any idea for above exception?

**Reply by dlemstra** [reply]
If you are running in `x64` I would advise you to use the `x64` version of the library instead of the `AnyCPU` version. Then you only need to make sure that the native library is also in the `bin` folder of your application.

**Reply by nsutariya9393** [reply]
should I manually copy native library from temp to bin folder? If so, I already tried it and it gives same error.

**Reply by dlemstra** [reply]
The temp folder will only be created when you use the `AnyCPU` version and I would recommend you to use the `x64` when you can. You can control the location where the native library will be stored with `MagickAnyCPU.CacheDirectory` if you want to keep using the `AnyCPU` version.

**Reply by nsutariya9393** [reply]
I changed to 64 and the exception is as below:
[DllNotFoundException: Unable to load DLL 'Magick.Native-Q16-x64.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)]
   ImageMagick.X64.Environment_Initialize() +0
   ImageMagick.Environment.Initialize() +122
   ImageMagick.NativeMagickSettings..cctor() +15

[TypeInitializationException: The type initializer for 'NativeMagickSettings' threw an exception.]
   ImageMagick.NativeMagickSettings..ctor() +0
   ImageMagick.MagickSettings..ctor() +94
   ImageMagick.MagickImage..ctor() +35
   ImageMagick.MagickImage..ctor(String fileName) +20
   PINT.btnSaveSKUPrintImages_Click(Object sender, EventArgs e) in c:\inetpub\wwwroot\test.aspx.cs:744
   System.Web.UI.WebControls.Button.OnClick(EventArgs e) +11750641
   System.Web.UI.WebControls.Button.RaisePostBackEvent(String eventArgument) +150
   System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint) +6016

**Reply by dlemstra** [reply]
Is the Native library in the bin folder?

**Reply by nsutariya9393** [reply]
No, should it be there? its not there for local too, but it works fine. Where should I get it?

**Reply by nsutariya9393** [reply]
so I copied Magick.Native-Q16-x64.dll from my local machine's temp folder to bin of my deployed folder, and it started giving me same error as before:
[DllNotFoundException: Unable to load DLL 'Magick.Native-Q16-x64.dll': The specified procedure could not be found. (Exception from HRESULT: 0x8007007F)]
   ImageMagick.X64.Environment_Initialize() +0
   ImageMagick.Environment.Initialize() +122
   ImageMagick.NativeMagickSettings..cctor() +15

[TypeInitializationException: The type initializer for 'NativeMagickSettings' threw an exception.]
   ImageMagick.NativeMagickSettings..ctor() +0
   ImageMagick.MagickSettings..ctor() +94
   ImageMagick.MagickImage..ctor() +35
   ImageMagick.MagickImage..ctor(String fileName) +20
   PINT.btnSaveSKUPrintImages_Click(Object sender, EventArgs e) in c:\inetpub\wwwroot\abc\PINT.aspx.cs:752
   System.Web.UI.WebControls.Button.OnClick(EventArgs e) +11750641
   System.Web.UI.WebControls.Button.RaisePostBackEvent(String eventArgument) +150
   System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint) +6016

**Reply by gh658804** [reply]
Hi. As i mentioned [here](https://github.com/dlemstra/Magick.NET/issues/366), i had just copied the **native** dll under syswow64 / system32 folders. I had also placed it in the bin folder along with everything else that belonged to that project. Have you tried all of these? Restarting IIS?

**Reply by nsutariya9393** [reply]
yes, I put native in all 3 folders and restarted IIS but didn't work. I wonder, is it something wrong with native? I copied it from my local. cause Local works fine just deployed doesn't work.

**Reply by gh658804** [reply]
I sympathize. I had faced such issues when i was battling with this problem. At the end it just worked.

What i did was to
1. Copy the native dll in all 3 folders
2. Make sure 32bit support (in my case, the site was 32-bit) was enabled in the app pool
3. I had created a specific app pool with a local user because i wanted it to have access to certain resources.

That's all i can remember at the moment. I do hope you figure it out.

**Reply by nsutariya9393** [reply]
thanks for your time. I did all those but none of them worked for me. 
I am not sure what you mean by specific app pool with a local user? 
also, can you tell me from where did you get native dll?

**Reply by gh658804** [reply]
I had downloaded it from the site, along with the main package. I made sure they were from the same version.

In the App pool of the site, in its advanced settings i had changed the identity to a local user of the server. I'm not sure if i had also set the "load user profile" to true and if it had played any role in this.

**Reply by nsutariya9393** [reply]
thank you for your help! I tried this too but no luck :-(

**Comment by gh658804** [comment]
Have in mind that the error i faced was the "The specified _module_ could not be found", not the "The specified _procedure_ could not be found" one you mentioned. This is a different issue which i don't think i can be of much help. One wild guess i can think off would be something like "the version of the native dll is different than the expected one".

In any case, i'm guessing that when the _module_ in the message turned into _procedure_, your program managed to find the native dll but failed for another reason.

I hope this will somehow help.

**Reply by nsutariya9393** [reply]
I was getting "The specified module could not be found" before and after adding native dll to bin folder, the error turned into "The specified procedure could not be found". One thing I would ask is, if you can send me the link to download native dll package from the website, I have downloaded it from nuget package manager console from visual studio.

**Reply by nsutariya9393** [reply]
yes my native dll and magick.net dll versions are different but not sure where I can get it. Help would be really appreciated.

**Reply by nsutariya9393** [reply]
I download a package and package version is 7.23.2.1 and the native inside is 7.0.11.2. I couldn't find same version native.

**Comment by gh658804** [comment]
Another thing i have stumbled upon several times is that the user running a .NET program must have write access to folder "C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files" (or the respective \Framework\ one for 32-bit apps. Perhaps you should try that?

**Reply by nsutariya9393** [reply]
This I already have. but as you said, something is wrong with native dll.
